New Polynomial Chaos Expansion using least squares page

mbaudin47 opened the pull request:
The goal of this PR is to provide a method to create a polynomial chaos expansion using least squares. This makes it possible to perform Sobol' analysis of the model using polynomial chaos expansion.

In particular, this PR provides a Treemap plot of the interaction Sobol' indices. This is an original plot, which has never been available elsewhere (except for internal projects at EDF). It is based on a specific feature available in OpenTURNS's PCE: we can get the interaction Sobol' indices of _any_ group of variables.

![image](https://github.com/user-attachments/assets/ab6b1e53-18bf-4397-aa66-f4b99180455b)

**Figure 1.** The Sobol' Treemap plot for the beam model using degree 4 least squares PCE and a sample size equal to 1024.

Moreover, this PR makes it possible to disable two existing analyses:

- Exploratory Analysis
- Expectation Analysis

Before this PR, these two analyses are enabled by default and the user cannot disable them. This can be useful in the cases where we are interested in other analyses.

MLegkovskis reviewed (commented):
### Review Comments

- **Local Testing**:  
  I performed a local test, and it looks great overall!

- **High-Dimensional Models (WIP)**:  
  I understand this is still a work in progress, but I have a question:  
  If a user selects, for example, the [[Morris Function](https://github.com/MLegkovskis/UncertaintyCat/blob/main/examples/Morris_Function.py)](https://github.com/MLegkovskis/UncertaintyCat/blob/main/examples/Morris_Function.py), which has 20 inputs, I would expect that the Polynomial Chaos Expansion (PCE) framework will struggle with such a high-dimensional problem.  
  On my local machine, it runs endlessly, and I assume the same issue would occur on the Snowflake servers hosting the Streamlit app.

- **Potential Solutions**:  
  1. Should we focus on integrating this OT-driven PCE framework into the [[PCE surrogate model module](https://github.com/MLegkovskis/UncertaintyCat/blob/main/pages/PCE_surrogate_model.py)](https://github.com/MLegkovskis/UncertaintyCat/blob/main/pages/PCE_surrogate_model.py), as you suggested in the forum? Or is this a completely separate stream of work?  
  2. Alternatively, we could introduce a dynamic mechanism to skip this module for models that are deemed too high-dimensional. What are your thoughts on this approach?

mbaudin47 commented:
Thank you for your analysis of the PCE on Morris. Actually, I had already extensively analysed the behaviour of this method on the Morris test case. It is relatively easy to get very satisfactory PCE expansion of the Morris function, because it is actually a polynomial. The relatively large number of inputs is not an issue, because only 10 variables are actually involved in the model. For example, here is the Treemap plot of the Morris model, using 1024 points and a non-sparse PCE. This leads to a 90% R2.

![image](https://github.com/user-attachments/assets/e2dce4ed-038f-4c5d-b665-74ca695e3a62)

It is easy to improve the accuracy of the PCE using a model selection method (such as LARS). The problem is then that the model selection algorithm can make the computation much longer: this is what happened in your example.

I think that tuning the algorithm may improve the behavior.

- [X] Disable the sparse mode by default to make it faster.
- [ ] Add a checkbutton to enable/disable the LARS model selection.
- [ ] Improve the search for insignificant group of variables so that the sum is lower than the threshold, instead of each individual interaction index as in the present implementation. 

Meanwhile, I set this PR as a draft. I will revert it to "ready for review" when it is ready.

mbaudin47 commented:
![image](https://github.com/user-attachments/assets/85eedf67-57bd-409f-bd24-2760b3ca18ac)

Figure 1. The Sobol' Treemap plot for the Morris model using 512 coefficients computed using least squares PCE and a sample size equal to 1024. The large number of rectangles is an indication of the complexity of the model. The top right rectangle represents less than 5% of output variance.

mbaudin47 commented:
I realize that I did not answer your questions.

- _"Should we focus on integrating this OT-driven PCE framework into the [[PCE surrogate model module](https://github.com/MLegkovskis/UncertaintyCat/blob/main/pages/PCE_surrogate_model.py)], as you suggested in the forum? Or is this a completely separate stream of work?"_ I do not know yet. From my point of view, this PR is some way to experiment with the app and Streamlit with outputs that cannot generated by other simple means. I suggest to hold this out until you feel sufficiently confident that it merges smoothly within the app. The current state of the PR is an inconsistent refactoring of the PCE surrogate module, with less feature.
- _"Alternatively, we could introduce a dynamic mechanism to skip this module for models that are deemed too high-dimensional. What are your thoughts on this approach?"_ For high dimensional models, the PR does not provide any consistent method because it does not provide a way to automatically select input variables. For this purpose, I developed another PCE-based tool using another combinatorial algorithm. But I do not think that I am ready to introduce it into the app, until the current PR is ready and merged.

mbaudin47 commented:
I feel that the current PR should be significantly improved in order to gets its way into the tool.

On the input side:
- Provide a graphical entry field to set the sample size as an integer.
- Provide a graphical checkbutton to enable the LARS method.
- Provide a graphical slider to set the basis size. This slider should be setup between 1 and the sample size - 1 in order to get a full rank design matrix. Actually, the code can produce a solution even if the basis size is larger than the sample size: 1) in full PCE mode, there is a regularization of the matrix, 2) in sparse PCE mode, the model selection can compute the significant coefficients only. But this sounds reasonable to me. 

On the output side:
- Export the PCE as a Python script, as discussed at https://openturns.discourse.group/t/introducing-an-ai-driven-uq-and-sensitivity-analysis-app-built-with-openturns/395/7

On the implementation:
- Fix the radial plot, as suggested in #20.

This would be a first usable draft of the tool which provides a PCE using least squares. Beyond this, the app could be extended it so as to support quadrature methods.

@MLegkovskis : How does this roadmap feels?

MLegkovskis commented:
Hi @mbaudin47,

Thank you for submitting this PR and for putting together the Sobol' analysis via Polynomial Chaos Expansion (PCE), along with the new Treemap visualization. 

### Sobol' Treemap and Unique Visualizations
First, I want to acknowledge the value this PR brings. The Treemap visualization of interaction Sobol' indices is genuinely unique and very cool looking, providing immediate insight. 

**Planned Steps:**
- **Adaptive Complexity Checks:**  
  We can implement checks that detect when a model’s complexity (dimensionality, polynomial degree, or sample size) surpasses practical thresholds for PCE. In such cases, the tool could:
  - Warn the user that PCE-based Sobol' analysis may be too costly and suggest using Monte Carlo-based indices.
  - Automatically skip PCE computation if certain criteria are met, providing a message to the user explaining why.
  - I have implemented something like this in the PCE-page of the app:
![image](https://github.com/user-attachments/assets/064edee3-4aea-4b7e-a744-012977ea77df)
 
- **Parameter Tuning Options:**  
  As you suggested, giving the user more control over parameters—like disabling LARS for sparse PCE or lowering the polynomial degree—can reduce complexity and computation time. This will make it easier to handle moderate-dimensional problems without endless runtime issues.

**Proposed Approach:**
- **Dedicated PCE Page Integration:**  
  Instead of making PCE-based Sobol' indices a standalone feature, could we incorporate it into the PCE surrogate module page that is driven off SALIB but make it work fully from OT. There, users will still be able to:
  - Build and validate a PCE model (checking R², residuals, etc.).
  - Tune parameters (e.g., sample size, polynomial degree, LARS toggle) in an environment designed for experimentation.
  - Once satisfied with the surrogate’s quality, they can seamlessly proceed to the Sobol' treemap visualization, which will act as a validation on its own, as sensible/physical Sobol' treemap validates your PCE surrogate in itself.

**Next Steps:**
1. **User Controls for PCE Input Parameters:**  
   - **Sample Size Input:** A field for users to specify the exact sample size to control complexity.
   - **LARS Enable/Disable Checkbutton:** A simple UI element to quickly toggle sparse model selection.
   - **Basis Size Slider:** Allow users to set the polynomial basis size, guiding them towards stable configurations and warning if the setting is outside recommended bounds.
   
2. **Output Improvements:**
   - **Exporting the PCE as a Script:**  
     As discussed, providing a way to export the PCE model as a Python script (or another portable format) will let users reuse the surrogate model within the main app.

### Summary and Outlook

Sorry if my response is a bit convoluted. Alongside your work on the Polynomial Chaos Expansion (PCE), I’ve also noted your efforts to improve the session state management within the Streamlit app and to fix the radial plots. If we could separate these improvements into distinct PRs—one for session state enhancements and another for radial plot fixes—it would be great should I understand that code correctly in terms of that is what it is trying to address?

To me, all of this great work suggests we should focus on refining the standalone SALIB-driven PCE page currently in development. I understand the challenges around copying and pasting raw polynomials into the main app due to their complexity + potential issues you have mentioned in the forum with this approach. Perhaps the best way forward is to direct further efforts into that dedicated PCE page, treating it as a self-contained environment for surrogate model creation and analysis, and making it full OT-driven to leverage your expertise.

If implementing a user workflow where the polynomial can be seamlessly integrated back into the main app proves too cumbersome, another possible approach would be to mirror the main app’s analyses within the PCE-based surrogate page itself. In other words, users could choose between a “direct” uncertainty quantification path—using their original model—and a “surrogate-driven” path—using the PCE-based model—without having to manually transfer polynomials. Streamlit’s flexibility should allow us to host both approaches within a single interface, clearly delineating between traditional methods and surrogate-driven methods for clarity. We can even make a page called just model analysis that tells the user --> ok go do PCE or stick with direct method?

Please let me know what you think of this direction.

MLegkovskis commented:
This is the page that I am talking about, just FYI:

![image](https://github.com/user-attachments/assets/2ff699e4-7d46-48f5-96a3-f52127e1742f)

MLegkovskis commented:
**Additional Thought:**  
I just had a light bulb moment—perhaps we don’t need to figure out how to copy and paste the surrogate into the main app at all. Instead, we could simply make a copy of `app.py` and call it something like `pce.py` or `surrogate.py`. So, the user would still input their models as usual, and on the backend, we’d generate the surrogate and run analyses identical to those in the main app, but powered by the surrogate instead of the direct model. This approach might significantly simplify the workflow, enabling users to decide whether to use direct or surrogate-based methods without the complexity of transferring polynomials. Let me know what you think!

mbaudin47 commented:
Thank you for your detailed message.

I am happy that you find some interest in the Treemap plot: new examples will show if that is an interesting plot and it's limitations as well.

_"If we could separate these improvements into distinct PRs—one for session state enhancements and another for radial plot fixes—it would be great._"

This is the good way to go indeed. Sorry for messing around that point in my PR: when I tried to make the PR as useful as possible, I realized that this could be improved.

mbaudin47 commented:
I reworked the PR and updated the following topics.
- Dedicated PCE Page Integration: ✓
- User Controls for PCE Input Parameters:
  - Sample Size Input: ✓
  - LARS Enable/Disable Checkbutton: ✓
  - Basis Size Slider: ✓

TODO:
- Merge the updated Radial Plot when available
- Exporting the PCE as a Script
- Plot the conditional expectation of the PCE

@MLegkovskis Before going further, please let me know if this fits your suggestions.

MLegkovskis commented:
> I reworked the PR and updated the following topics.
> 
> * Dedicated PCE Page Integration: ✓
> * User Controls for PCE Input Parameters:
>   
>   * Sample Size Input: ✓
>   * LARS Enable/Disable Checkbutton: ✓
>   * Basis Size Slider: ✓
>   * Restore the default settings in `app.py`: ✓
> 
> ![image](https://private-user-images.githubusercontent.com/31351465/398265055-1ba24a35-0374-44ec-b622-612f4ed3863e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU4NDE2NDEsIm5iZiI6MTczNTg0MTM0MSwicGF0aCI6Ii8zMTM1MTQ2NS8zOTgyNjUwNTUtMWJhMjRhMzUtMDM3NC00NGVjLWI2MjItNjEyZjRlZDM4NjNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTAyVDE4MDkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFlZWI5N2ZlMmJlZmUwZjRkMWY2Y2UzMzg0YTVlMTNjOWUzMzM5YjdmYThkYTk0Y2IwMWFjNjQ4YjRmMjQ0ODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.oC0wok5GiHSb1O2CxZxFCZlgAUE7eGMtek7bc0qkcSY) **Figure 1.** The new settings of the OpenTURNS's based PCE.
> 
> **TODO:**
> 
> * Merge the updated Radial Plot when PR [Improved Sobol' radial plot. #23](https://github.com/MLegkovskis/UncertaintyCat/pull/23) is available
> * Exporting the PCE as a Script
> 
> **Perspectives**
> 
> * Plot the conditional expectation of the PCE (see [plot_chaos_conditional_expectation](http://openturns.github.io/openturns/latest/auto_meta_modeling/polynomial_chaos_metamodel/plot_chaos_conditional_expectation.html?highlight=conditional+expectation))
> * Use other designs of experiments: 1) Monte-Carlo (the default), 2) Quasi Monte-Carlo, 3) Latin Hypercube Sampling (LHS)
> 
> @MLegkovskis Before going further, please let me know if this fits your suggestions.
> 
> Edit (27/12/2024). I integrated the export feature. It does not work for some distributions including the log-normal's parametrized distribution: see [2858](https://github.com/openturns/openturns/issues/2858). This prevents to create the metamodel for the beam model. For this particular model, the export is:
> 
> ```python
> marginals.append(ParametrizedDistribution(30000.0, 9000.0, 15000.0))
> ```
> 
> The corrected code is:
> 
> ```python
> marginals.append(ParametrizedDistribution(LogNormalMuSigma(30000.0, 9000.0, 15000.0)))
> ```

Hi Michael, first of all, thank you for this great work! I’ve done some local testing, and everything looks good on my end. Given your expertise with polynomial chaos expansions, I agree we should avoid directly copying raw polynomials into the main application (as that can introduce complications) + use openturns, of course. Instead, I propose that the new Least Squares PCE page effectively *becomes* our main route for building polynomial chaos expansions and performing the associated surrogate-based uncertainty analysis.  

Rather than trying to integrate a PCE definition back into the existing app (by means of copy-paste), I’d suggest duplicating the same modules and functionality from `app.py`, but having them run off the PCE surrogate under the hood instead of the “raw” model (we can introduce a new module just simply telling the user this is the originally-supplied model and this is PCE version, and let AI comment on the accuracy just for fun/user-friendliness). Essentially, the new page would replicate the standard analyses (e.g., Sobol’ indices, treemap, radial plots, Taylor, Coeffs even Shapley will work fine etc) by using `metamodel` (i.e., our PCE surrogate) wherever the main app currently uses the original function. 

The quick pseudocode below illustrates that idea—once `chaosResult` is computed, you can create `metamodel = chaosResult.getMetaModel()`, then feed it into an OpenTURNS PythonFunction just as the main app does with the raw model. This way, the user can choose to do a “surrogate-driven” analysis without having to copy/paste polynomials into the main app.  

So basically, PCE_least_squares.py = app.py but "simply" we feed into all the modules: `ot_model = ot.PythonFunction(problem['num_vars'], 1, metamodel` instead of the current `ot_model = ot.PythonFunction(problem['num_vars'], 1, model)`

Let me know what you think, and thank you again for pushing this forward!

mbaudin47 commented:
You are right: exporting the surrogate model into a string is much more complicated than re-using it. Let me rephrase your suggestion as follows. We ask the user a surrogate metamodel name and use it to reference it as a new physical model. The procedure would be:
1. In the "PCE from least squares" page, select a physical model. Estimate it's PCE coefficients. Choose a name and export the PCE. This creates a new physical model with the given name and the same probabilistic model.
2. In the main app, pick this model from the list and propagate the uncertainties through it as a regular model.

Is that OK from your point of view?

MLegkovskis commented:
hi @mbaudin47, I am proposing **two separate, standalone pages** for uncertainty quantification:

1. **UQ via Direct Model**  
   - The user inputs their direct model, and the page performs all uncertainty analyses (Sobol indices, Taylor expansions, etc.).  
   - This corresponds to the current `app.py` file but will be renamed to something like `uq_direct.py`.

2. **UQ via PC**  
   - Under the hood, this page performs **essentially the same analyses as `uq_direct.py`** but leverages a surrogate (PCE) automatically generated in the backend (no user “export” steps, etc.).  
   - We’ll simply provide a basic view of the surrogate (e.g., coefficients) for transparency as an extra module, but there’s no manual copy–paste or “jumping” between pages.  
   - You stay on one of these pages throughout the entire UQ analysis—either on the `uq_direct` page or the `uq_pce` page. The two pages are nearly identical in terms of the underlying Python code; the only difference is that `uq_pce` will also create a PCE surrogate of the user-entered model. 


Please let me know if this makes sense?

mbaudin47 commented:
Thank you for the detailed road map. Your point of view is clearer now. I think that this option has several drawbacks.
- This duplicates the features from the user's point of view. As a consequence, we might get confused on the results we are analysing.
- The user may not fully understand that the validation of the surrogate must be done **before** using it. Hence, some users may jump to the results without noticing that e.g. the R2 validation score is not satisfactory.
- It might produce a too long page, diminishing the user's ability to read the results.
- This makes it difficult to extend to other features. For example, if we want to provide a method to estimate the probability that the output Y exceeds a given threshold, then another pair of pages has to be created : 1) direct, 2) from surrogate. 

On the other hand, it also has some advantages. 
- It simplifies the user's actions on the app. The other method requires to use 2 different pages, which can be cumbersome for some users. 

I guess I have to think about it twice before jumping to a conclusion because  it might produce a disappointing result.

MLegkovskis commented:
Hi Michael,

Thank you for laying out your thoughts so clearly—your points about validation and avoiding a flood of results without first confirming the surrogate’s accuracy really resonate with me. I totally agree that simply offering two parallel pages (“UQ via Direct Model” and “UQ via Surrogate”) could encourage users to dive straight into a surrogate-based analysis without checking whether \(R^2\) or other metrics are acceptable. It also risks duplicating every feature we add down the line, effectively doubling our maintenance overhead. Those are all legitimate concerns.

Regarding your idea to export and import a PCE model (or any model, really) as a Python/XML file: I think that’s a fantastic approach. It preserves user flexibility—someone can build and validate a PCE on one machine, save the file, then load it later (or share it with collaborators), all without us having to jam everything into a single monolithic page or maintain additional dynamic lists etc. We can easily add a “Download Surrogate” button on the “PCE from least squares” page and then provide a straightforward “Upload Model” functionality on the main UQ page. This way, we avoid the copy-paste pitfalls, and users can manage their own library of surrogates or raw models on their local disk.

I’m on board with this direction. It neatly solves the issues we’ve both highlighted: minimal duplication of analysis features, more intuitive validation (since the user must explicitly decide to export a “good” surrogate), and a smoother way to scale as more advanced functionalities come into play. Thank you again for driving this.

mbaudin47 commented:
@MLegkovskis : In d9003c0, I modified the PCE page to add a new Download button. In the main app, I created a new column and created a new Upload button to upload the PCE surrogate.

![image](https://github.com/user-attachments/assets/9fd0b774-41b6-47c9-b9e3-30f4d4b87376)

**Figure 1.** The new Download button in the "PCE least squares" page.

![image](https://github.com/user-attachments/assets/1222c520-5363-408f-bb49-48fd9252296f)

**Figure 2.** The new Upload button in the main app page.

As a first attempt, I downloaded and uploaded the Python script (although we know that there are limitations for the LogNormalMuSigma distribution). I was not able to make the main app to be updated correctly when I upload the file: it seems that the action is ignored by the tool. I think that I do not understand how StreamLit works here. Can you please have a look and give some suggestions?

MLegkovskis commented:
> @MLegkovskis : In [d9003c0](https://github.com/MLegkovskis/UncertaintyCat/commit/d9003c09ac54d5d2a19359d3a6be6cc854ee86a7), I modified the PCE page to add a new Download button. In the main app, I created a new column and created a new Upload button to upload the PCE surrogate.
> 
> ![image](https://private-user-images.githubusercontent.com/31351465/401713859-9fd0b774-41b6-47c9-b9e3-30f4d4b87376.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY2NzE4NzUsIm5iZiI6MTczNjY3MTU3NSwicGF0aCI6Ii8zMTM1MTQ2NS80MDE3MTM4NTktOWZkMGI3NzQtNDFiNi00N2M5LWI5ZTMtMzBmNGQ0Yjg3Mzc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEyVDA4NDYxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ5NTc5NzZmNDk1YzY4OTg3YzgyMzBlMGI2MDU4ZTU0YjU2ODk0MDVkMjMwZGU1OGQ5NWU4ZjI1MjFlMTI4YmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-_jBkW_nYfGsOekOiPWlEvmLpsF5_HHZePXEKWwiBFA)
> 
> **Figure 1.** The new Download button in the "PCE least squares" page.
> 
> ![image](https://private-user-images.githubusercontent.com/31351465/401712919-1222c520-5363-408f-bb49-48fd9252296f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY2NzE4NzUsIm5iZiI6MTczNjY3MTU3NSwicGF0aCI6Ii8zMTM1MTQ2NS80MDE3MTI5MTktMTIyMmM1MjAtNTM2My00MDhmLWJiNDktNDhmZDkyNTIyOTZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEyVDA4NDYxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg2OTJjMTIzZTYzNmRmN2EyMDcwYTMyZWNmM2E4YmY4ZTE3MjEwOWM5Njk2YTU3MDMxNThjMjgzMWFiNDNlMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.e_ZXHJPq5Wo2CKDjBUHfRKSeEYmJidrYG7gDYRSZp90)
> 
> **Figure 2.** The new Upload button in the main app page.
> 
> As a first attempt, I downloaded and uploaded the Python script (although we know that there are limitations for the LogNormalMuSigma distribution). I was not able to make the main app to be updated correctly when I upload the file: it seems that the action is ignored by the tool. I think that I do not understand how StreamLit works here. Can you please have a look and give some suggestions?

Please take a look: https://github.com/MLegkovskis/UncertaintyCat/pull/32

mbaudin47 commented:
Thank you very much!
I will look at his more deeply and let you know my point of view. 

> > @MLegkovskis : In [d9003c0](https://github.com/MLegkovskis/UncertaintyCat/commit/d9003c09ac54d5d2a19359d3a6be6cc854ee86a7), I modified the PCE page to add a new Download button. In the main app, I created a new column and created a new Upload button to upload the PCE surrogate.
> > 
> > ![image](https://private-user-images.githubusercontent.com/31351465/401713859-9fd0b774-41b6-47c9-b9e3-30f4d4b87376.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY2NzE4NzUsIm5iZiI6MTczNjY3MTU3NSwicGF0aCI6Ii8zMTM1MTQ2NS80MDE3MTM4NTktOWZkMGI3NzQtNDFiNi00N2M5LWI5ZTMtMzBmNGQ0Yjg3Mzc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEyVDA4NDYxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ5NTc5NzZmNDk1YzY4OTg3YzgyMzBlMGI2MDU4ZTU0YjU2ODk0MDVkMjMwZGU1OGQ5NWU4ZjI1MjFlMTI4YmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-_jBkW_nYfGsOekOiPWlEvmLpsF5_HHZePXEKWwiBFA)
> > 
> > **Figure 1.** The new Download button in the "PCE least squares" page.
> > 
> > ![image](https://private-user-images.githubusercontent.com/31351465/401712919-1222c520-5363-408f-bb49-48fd9252296f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY2NzE4NzUsIm5iZiI6MTczNjY3MTU3NSwicGF0aCI6Ii8zMTM1MTQ2NS80MDE3MTI5MTktMTIyMmM1MjAtNTM2My00MDhmLWJiNDktNDhmZDkyNTIyOTZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEyVDA4NDYxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg2OTJjMTIzZTYzNmRmN2EyMDcwYTMyZWNmM2E4YmY4ZTE3MjEwOWM5Njk2YTU3MDMxNThjMjgzMWFiNDNlMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.e_ZXHJPq5Wo2CKDjBUHfRKSeEYmJidrYG7gDYRSZp90)
> > 
> > **Figure 2.** The new Upload button in the main app page.
> > 
> > As a first attempt, I downloaded and uploaded the Python script (although we know that there are limitations for the LogNormalMuSigma distribution). I was not able to make the main app to be updated correctly when I upload the file: it seems that the action is ignored by the tool. I think that I do not understand how StreamLit works here. Can you please have a look and give some suggestions?
> 
> Please take a look: https://github.com/MLegkovskis/UncertaintyCat/pull/32

MLegkovskis commented:
hi @mbaudin47  I have carried out all the testing (hopefully) and I think this is ready to go.

mbaudin47 commented:
Sorry for the delay. I have group of students in my course during January and February and I have much less time  available for this project. I will review this as soon as possible. 
By the way, can you please send me a private message on michael. baudin@gmail.com?

mbaudin47 commented:
Great news! Thank you very much for this work.
